atc/db/lock/lock.go

```diff
@@ -29,7 +30,7 @@ const (
 var ErrLostLock = errors.New("lock was lost while held, possibly due to connection breakage")
 
 func NewBuildTrackingLockID(buildID int) LockID {
-	return LockID{LockTypeBuildTracking, buildID}
+	return LockID{LockTypeBuildTracking, buildID % MaxInt}
```

Comment:
Like I suggested in https://github.com/concourse/concourse/issues/8364#issuecomment-1141557700, I still think the change should be in `LockID.toDBArgs()`. This is not just a problem with build id, I would consider volume id, container id to be larger than build id. As the restriction comes from PG, the fix should be as close as possible to db related code.

If you change `LockID.toDBArgs()`, please add some tests accordingly.

Reply:
For hush-house (being up for 4 years), we have the maximum ids as following
```
1081477231 build id
214934158  container id
410257222  volumn id
```

Could you list the ids in runway for comparation? Ideally I want to keep the impact as minimum as possible, until we see the same problem reported for container/volumn ID.

Reply:
Our prod cluster has been running for ~3 years. To roughly calculate, say 1000 days, `10422031 / 1000 = 10422` builds per day, that's a reasonable number. As most of builds are triggered by code check-in. 

Remember that, 6.x uses `checks` table, so that checks doesn't consume any build id. 

From the other perspective, majority workloads on our cluster are coming from checks, that's reason why I spent so much efforts on optimizing checks.

Reply:
I don't really want to add this overhead dividend operation to not needed lock ids so to me LockID.toDBArgs() is the least place I want to solve the problem. Plus the way you suggested will also calculate the lock_type/maxint32 which is totally a waste.

 I agree that volume id and container id should be considered though. However it seems there is no lock ID that depends on container ID.

Reply:
Comparing to the following DB operation `pg_try_advisory_lock`, cost of `%` operator can be ignored.

With your current implementation, there are still potential problems:

* When container_id related lock is added, or other lock types that need build id or volume id are added, future developers have to remember `% MaxInt32`, which is error-prone.
* Job id, resource configure id may eventually reach to MaxInt32 though possibility is low.

The following implementation will not `%` lock type:

```golang
func (l LockID) toDBArgs() []interface{} {
	result := []interface{}{l[0]}
	if len(l) > 1 {
             result = append(result, l[1] % math.MaxInt32
        }

	return result
}
```

Reply:
> * When container_id related lock is added, or other lock types that need build id or volume id are added, future developers have to remember `% MaxInt32`, which is error-prone.

I have added a block of comment for furture reference of developing lock type. Plus I can hardly imagine there will be a change of existing lock type or adding a new one.

Reply:
TBH, I am not convinced. I still think the change should be in `toDBArgs()`. But anyway I don't want to block you.